lib/publiccloud/instance.pm

```diff
@@ -85,6 +86,43 @@ sub run_ssh_command {
     }
 }
 
+=head2 ssh_script_retry
+
+    ssh_script_retry(command[, retry => 3][, delay => 10][, timeout => 90][, ssh_opts =>'..'][, username => 'XXX'][, no_quote => 0]);
+
+Run a C<command> via ssh in the given PC instance until it succeeds or
+the given number of retries is exhausted and an exception is thrown.
+Timeout can be set by C<timeout> or 90 sec by default.
+By default, the command is passed in single quotes to SSH.
+To avoid quoting use C<<no_quote=>1>>.
+With C<<ssh_opts=>'...'>> you can overwrite all default ops which are in
+C<<$instance->ssh_opts>>.
+Use argument C<username> to specify a different username then
+C<<$instance->username()>>.
+=cut
+sub ssh_script_retry {
+    my ($self, $cmd, %args) = @_;
+    $cmd = $args{cmd} if (!defined $cmd && $args{cmd});
+    $args{rc_only} = 1;
+    my $tries = $args{retry} // 3;
+    my $delay = $args{delay} // 10;
+
+    for (my $try = 0; $try < $tries; $try++) {
+        my $rc = 1;
+        # ssh_script_run can throw a timeout exception
+        eval {
+            $rc = $self->run_ssh_command($cmd, %args);
+        };
+        if ($@ or $rc != 0) {
+            sleep($delay);
+            next;
+        } else {
+            return $rc;
+        }
```

Comment:
```suggestion
        my $rc = $self->run_ssh_command($cmd, %args);
        return $rc if ($rc != 0);
        sleep($delay);
```

Reply:
you would not catch SSH timeout failures... but maybe we do not want to

Reply:
There should be
```perl
return $rc if ($rc == 0);
```

Reply:
ssh timeouts are one of the reasons why we are doing this - I could even see some in the verification runs.
This is the reason for the whole `eval` block, which indeed makes the code a bit messy here

Reply:
Wouldn't the suggestion I placed prevent SSH timeouts? The `return $rc if ($rc == 0);` will not fail either return but in case of SSH timeout the `for` cycle will continue.

Reply:
And btw. if there is a timeout, we should increase it.

Reply:
Timeouts are build-in into `run_ssh_command` - See https://github.com/grisu48/os-autoinst-distri-opensuse/blob/9c365be6d60a4426f24fdcb41acce20d7fefe03f/lib/publiccloud/instance.pm#L55
If the `timeout` setting is not set, it defaults to 90, which should be sufficient for most tasks (we could consider increasing this for SCC though)
Setting the `timeout` to zero will run the command without waiting and returning the return code, so that's also not an option.

In other words: If we want to keep `run_ssh_command` as it is, the only way I see is to set `rc_only`, which requires a timeout and will throw an expection on timeout. We need to account for this with the `eval` block.

Reply:
The [lib/publiccloud/instance.pm#L81](https://github.com/grisu48/os-autoinst-distri-opensuse/blob/9c365be6d60a4426f24fdcb41acce20d7fefe03f/lib/publiccloud/instance.pm#L81) should not happen in this loop. There should be something like `die => 0` option for the `run_ssh_command` routine.

Reply:
Included